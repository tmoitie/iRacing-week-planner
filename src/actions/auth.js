import {
  getAuth, signOut as fbSignOut, signInWithEmailAndPassword, createUserWithEmailAndPassword,
  onAuthStateChanged, sendPasswordResetEmail, GoogleAuthProvider, signInWithPopup,
} from 'firebase/auth';
import { debouncedDispatcherSaveSettings, getSettingsFromFirebase, saveSettingsToFirebase } from './settings';

export const LOADING_AUTH = 'AUTH/LOADING_SIGN_IN';
export const ERROR_AUTH = 'AUTH/ERROR_AUTH';
export const ERROR_RESET = 'AUTH/ERROR_RESET';
export const ERROR_ACKNOWLEDGE = 'AUTH/ERROR_ACKNOWLEDGE';
export const SIGNED_IN = 'AUTH/SIGNED_IN';
export const SIGNED_OUT = 'AUTH/SIGNED_OUT';
export const RESET_SENT = 'AUTH/RESET_SENT';
export const LOADING_RESET = 'AUTH/LOADING_RESET';

export function signOut() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    if (getState().settings.firebaseSynced === false) {
      debouncedDispatcherSaveSettings.cancel();
      await dispatch(saveSettingsToFirebase());
    }
    const auth = getAuth(getState().auth.firebaseApp);
    await fbSignOut(auth);

    dispatch({ type: SIGNED_OUT });
  };
}

export function signIn(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await signInWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function signInWithGoogle() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);

      const provider = new GoogleAuthProvider();
      provider.addScope('profile');
      provider.addScope('email');

      await signInWithPopup(auth, provider);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function createAccount(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await createUserWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function acknowledgeAuthError() {
  return { type: ERROR_ACKNOWLEDGE };
}

export function signedIn(user) {
  return async (dispatch) => {
    await dispatch({ type: SIGNED_IN, user });

    if (user) {
      await dispatch(getSettingsFromFirebase());
    }
  };
}

export function startListener() {
  return async (dispatch, getState) => {
    const auth = getAuth(getState().auth.firebaseApp);
    const { currentUser } = auth;
    dispatch(signedIn(currentUser));

    onAuthStateChanged(auth, (user) => {
      dispatch(signedIn(user));
    });
  };
}

export function forgottenPassword(email) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_RESET });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await sendPasswordResetEmail(auth, email);
      dispatch({ type: RESET_SENT });
      return { type: RESET_SENT };
    } catch (error) {
      dispatch({ type: ERROR_RESET, error });
      return { type: ERROR_RESET };
    }
  };
}
