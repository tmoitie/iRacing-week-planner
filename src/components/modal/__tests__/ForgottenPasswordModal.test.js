import { describe, test } from '@jest/globals';
import React from 'react';
import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { getAuth, sendPasswordResetEmail } from 'firebase/auth';

import ForgottenPasswordModal from '../ForgottenPasswordModal';
import { ERROR_ACKNOWLEDGE } from '../../../actions/auth';

describe('components/modal/ForgottenPasswordModal', () => {
  const mockStore = configureMockStore([thunk]);

  test('renders closed', async () => {
    const store = mockStore({ auth: { errorReset: null, loadingReset: false } });
    const { baseElement } = render(<Provider store={store}><ForgottenPasswordModal isOpen={false} /></Provider>);
    expect(screen.queryByTestId('forgottenPasswordModal')).not.toBeTruthy();
  });

  test('happy path', async () => {
    const user = userEvent.setup();

    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    const store = mockStore({ auth: { errorReset: null, loadingReset: false, firebaseApp: {} } });
    const { baseElement, unmount } = render(<Provider store={store}><ForgottenPasswordModal isOpen /></Provider>);

    expect(await screen.findByTestId('forgottenPasswordModal')).toMatchSnapshot();

    await user.type(screen.getByRole('textbox', { name: 'Email address' }), 'tom@example.com');
    await user.click(screen.getByText('Submit'));

    expect(sendPasswordResetEmail).toHaveBeenCalledWith(firebaseAuth, 'tom@example.com');

    expect(await screen.findByTestId('forgottenPasswordModal')).toMatchSnapshot();

    const thanksModal = await screen.findByTestId('forgottenPasswordThanks');
    await user.click(within(thanksModal).getByText('Close'));

    expect(await screen.findByTestId('forgottenPasswordModal')).toMatchSnapshot();
  });

  test('reset error', async () => {
    const user = userEvent.setup();

    const firebaseAuth = {};
    getAuth.mockReturnValue(firebaseAuth);
    sendPasswordResetEmail.mockRejectedValueOnce({ message: 'error' });

    const store = mockStore({ auth: { errorReset: null, loadingReset: false, firebaseApp: {} } });
    const { baseElement } = render(<Provider store={store}><ForgottenPasswordModal isOpen /></Provider>);

    await user.click(screen.getByText('Submit'));

    expect(await screen.findByTestId('forgottenPasswordModal')).toMatchSnapshot();
  });

  test('renders error', async () => {
    const user = userEvent.setup();

    const store = mockStore({ auth: { errorReset: { message: 'Error message' }, loadingReset: false } });
    const dispatchSpy = jest.spyOn(store, 'dispatch');
    const { baseElement } = render(<Provider store={store}><ForgottenPasswordModal isOpen /></Provider>);

    expect(await screen.findByTestId('forgottenPasswordModal')).toMatchSnapshot();

    await user.click(within(await screen.findByRole('alert')).getByLabelText('Close'));
    expect(dispatchSpy).toHaveBeenCalledWith({ type: ERROR_ACKNOWLEDGE });
  });

  test('renders loading', async () => {
    const store = mockStore({ auth: { errorReset: null, loadingReset: true } });
    const { baseElement } = render(<Provider store={store}><ForgottenPasswordModal isOpen /></Provider>);

    expect(await screen.findByTestId('forgottenPasswordModal')).toMatchSnapshot();
  });
});
