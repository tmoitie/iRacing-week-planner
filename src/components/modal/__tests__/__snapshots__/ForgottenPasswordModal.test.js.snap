// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/modal/ForgottenPasswordModal happy path 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="forgottenPasswordModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Reset Password
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Enter your email address to reset your password.
        </p>
        <form>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <button
            class="btn btn-default"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal happy path 2`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="forgottenPasswordModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Reset Password
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Enter your email address to reset your password.
        </p>
        <form>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <button
            class="btn btn-default"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal happy path 3`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="forgottenPasswordModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Reset Password
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Enter your email address to reset your password.
        </p>
        <form>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <button
            class="btn btn-default"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal renders error 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="forgottenPasswordModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Reset Password
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Enter your email address to reset your password.
        </p>
        <form>
          <div
            class="alert alert-warning alert-dismissible"
            role="alert"
          >
            <button
              aria-label="Close"
              class="close"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
            Error message
          </div>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <button
            class="btn btn-default"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal renders loading 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="forgottenPasswordModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Reset Password
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Enter your email address to reset your password.
        </p>
        <div>
          Loading
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/modal/ForgottenPasswordModal reset error 1`] = `
<div
  class="Modal__Bootstrap modal-dialog modal-lg"
  data-testid="forgottenPasswordModal"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          class="glyphicon glyphicon-remove"
        />
      </button>
      <h4
        class="modal-title"
      >
        Reset Password
      </h4>
    </div>
    <div
      class="modal-body"
      style="max-height: 55vh; overflow-y: auto;"
    >
      <div
        class="container-fluid"
      >
        <p>
          Enter your email address to reset your password.
        </p>
        <form>
          <div
            class="form-group"
          >
            <label
              for="loginEmail"
            >
              Email address
            </label>
            <input
              class="form-control"
              id="loginEmail"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <button
            class="btn btn-default"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
